### PR TITLE
Storages: load RSResult only once

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #6233

Problem Summary:

### What is changed and how it workss

```commit-message
Before, TiFlash will load the `RSResult` three times for one query (MVCC/Build Bitmap/Query).

This PR introduces `DMFilePackFilterResult` which load `RSResult` only once, and only passes the `RSResult` to `DMFileReader`.
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 1]
```diff
@@ -3525,6 +3526,7 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(
             read_data_block_rows);
     }
 
+    std::cout << "getBitmapFilterInputStream" << std::endl;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Segment.cpp]:
```suggestion
```

**Review by CalvinNeo — APPROVED** [review]
lgtm

File: dbms/src/Storages/DeltaMerge/StableValueSpace.cpp [hunk 2]
```diff
@@ -461,11 +450,69 @@ SkippableBlockInputStreamPtr StableValueSpace::Snapshot::getInputStream(
...
+    else
+    {
+        return std::make_shared<ConcatSkippableBlockInputStream</*need_row_id*/ false>>(
+            streams,
+            std::move(rows),
+            context.scan_context);
+    }
+}
+
+SkippableBlockInputStreamPtr StableValueSpace::Snapshot::tryGetInputStreamWithVectorIndex(
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/StableValueSpace.cpp:506]: Why do we need to add this method? Seems it is the same as `StableValueSpace::Snapshot::getInputStream` except ann_query_info is not nullptr?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/StableValueSpace.cpp:504]: I will continue to separate all `tryGetInputStreamWithVectorIndex` in the next PR, to make it more clear.

File: dbms/src/Storages/DeltaMerge/File/DMFilePackFilterResult.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    void tryLoadIndex(ColId col_id) const;
+
+private:
+    MinMaxIndexCachePtr index_cache;
+    FileProviderPtr file_provider;
+    ReadLimiterPtr read_limiter;
+
+    const ScanContextPtr scan_context;
+
+    DMFilePtr dmfile;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFilePackFilterResult.h:88]: checkout this PR: https://github.com/Lloyd-Pottiger/tiflash/pull/22

I think it is dangerous and not necessary to keep a shared_ptr to DMFile in `DMFilePackFilterResult`. Let's 

* Replace `DMFilePackFilterResult::validRowsAndBytes` by `DMFilePackFilter::loadValidRowsAndBytes`
* Remove shared_ptr to DMFile and other useless member from `DMFilePackFilterResult`

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**